In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

# Adjust this path if your project structure differs
DATA_PATH = Path('../data/raw/metabric.csv')

df = pd.read_csv(DATA_PATH)
print(f"Shape: {df.shape}")
print(f"\nColumn names ({len(df.columns)}):")
for col in df.columns:
    print(f"  {col!r}")

Shape: (2509, 40)

Column names (40):
  'Patient ID'
  'Age at Diagnosis'
  'Type of Breast Surgery'
  'Cancer Type'
  'Cancer Type Detailed'
  'Cellularity'
  'Chemotherapy'
  'Pam50 + Claudin-low subtype'
  'Cohort'
  'ER status measured by IHC'
  'ER Status'
  'G-Neoplasm Histologic Grade'
  'HER2 status measured by SNP6'
  'HER2 Status'
  'Tumor Other Histologic Subtype'
  'Hormone Therapy'
  'Inferred Menopausal State'
  'Integrative Cluster'
  'Primary Tumor Laterality'
  'Lymph nodes examined positive'
  'Mutation Count'
  'Nottingham prognostic index'
  'Oncotree Code'
  'Overall Survival (Months)'
  'Overall Survival Status'
  'PR Status'
  'Radio Therapy'
  'Relapse Free Status (Months)'
  'Relapse Free Status'
  'Sex'
  '3-Gene classifier subtype'
  'Tumor Size(mm)'
  'Tumor Stage'
  "Patient's Vital Status"
  'TNBC'
  'S-Tumor-Size'
  'NPI-Prognosis'
  'Survival Status'
  'RecurredNotRecurredStatus'
  'Age40'


In [2]:
# Dtypes and missingness in one table
inventory = pd.DataFrame({
    'dtype': df.dtypes,
    'non_null': df.notna().sum(),
    'null_count': df.isna().sum(),
    'null_pct': (df.isna().sum() / len(df) * 100).round(2),
    'n_unique': df.nunique(),
})
inventory = inventory.sort_values('null_pct', ascending=False)
print(inventory.to_string())

                                  dtype  non_null  null_count  null_pct  n_unique
3-Gene classifier subtype           str      1764         745     29.69         4
Tumor Stage                     float64      1788         721     28.74         5
Primary Tumor Laterality            str      1870         639     25.47         2
Cellularity                         str      1917         592     23.60         3
Type of Breast Surgery              str      1955         554     22.08         2
HER2 Status                         str      1980         529     21.08         2
Chemotherapy                        str      1980         529     21.08         2
Pam50 + Claudin-low subtype         str      1980         529     21.08         7
PR Status                           str      1980         529     21.08         2
Radio Therapy                       str      1980         529     21.08         2
Patient's Vital Status              str      1980         529     21.08         3
Hormone Therapy 

In [3]:
# For every object-typed column, dump unique values with counts
categorical_cols = df.select_dtypes(include=['object']).columns.tolist()

for col in categorical_cols:
    print(f"\n{'='*60}")
    print(f"  {col}")
    print('='*60)
    vc = df[col].value_counts(dropna=False)
    print(vc.to_string())


  Patient ID
Patient ID
MB-0000      1
MB-0002      1
MB-0005      1
MB-0006      1
MB-0008      1
MB-0010      1
MB-0014      1
MB-0020      1
MB-0022      1
MB-0025      1
MB-0028      1
MB-0035      1
MB-0036      1
MB-0039      1
MB-0045      1
MB-0046      1
MB-0048      1
MB-0050      1
MB-0053      1
MB-0054      1
MB-0056      1
MB-0059      1
MB-0060      1
MB-0062      1
MB-0064      1
MB-0066      1
MB-0068      1
MB-0071      1
MB-0079      1
MB-0081      1
MB-0083      1
MB-0093      1
MB-0095      1
MB-0097      1
MB-0099      1
MB-0100      1
MB-0101      1
MB-0102      1
MB-0106      1
MB-0107      1
MB-0108      1
MB-0109      1
MB-0110      1
MB-0111      1
MB-0112      1
MB-0113      1
MB-0114      1
MB-0115      1
MB-0116      1
MB-0117      1
MB-0119      1
MB-0120      1
MB-0121      1
MB-0122      1
MB-0123      1
MB-0124      1
MB-0125      1
MB-0126      1
MB-0127      1
MB-0128      1
MB-0129      1
MB-0130      1
MB-0131      1
MB-0133      1
MB-0134      1


C:\Users\liamt\AppData\Local\Temp\ipykernel_4172\2685467197.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = df.select_dtypes(include=['object']).columns.tolist()


In [4]:
# Summary stats for every numeric column
numeric_cols = df.select_dtypes(include=['int64', 'float64']).columns.tolist()

print(f"Numeric columns ({len(numeric_cols)}):")
print(df[numeric_cols].describe().T.to_string())

Numeric columns (13):
                                count        mean        std    min        25%         50%         75%     max
Age at Diagnosis               2498.0   60.420300  13.032997  21.93  50.920000   61.110000   70.000000   96.29
Cohort                         2498.0    2.900320   1.962216   1.00   1.000000    3.000000    4.000000    9.00
G-Neoplasm Histologic Grade    2388.0    2.412060   0.649363   1.00   2.000000    3.000000    3.000000    3.00
Lymph nodes examined positive  2243.0    1.950513   4.017774   0.00   0.000000    0.000000    2.000000   45.00
Mutation Count                 2357.0    5.578702   3.967967   1.00   3.000000    5.000000    7.000000   80.00
Nottingham prognostic index    2287.0    4.028787   1.189092   1.00   3.048000    4.044000    5.040000    7.20
Overall Survival (Months)      1981.0  125.244271  76.111772   0.00  60.866667  116.466667  185.133333  355.20
Relapse Free Status (Months)   2388.0  108.842487  76.519494   0.00  40.560000   99.095000

In [5]:
# Targeted cross-checks — what does the PDF say each value should be?

# ER Status: should be Positive/Negative (no typos)
print("ER Status values:")
if 'ER Status' in df.columns:
    print(df['ER Status'].value_counts(dropna=False))
print()

# ER status measured by IHC: known to have "Positve" typo in week 1 file
print("ER status measured by IHC values:")
if 'ER status measured by IHC' in df.columns:
    print(df['ER status measured by IHC'].value_counts(dropna=False))
print()

# Cellularity: should be High/Moderate/Low (PDF says "high cellularity is worse")
print("Cellularity values:")
if 'Cellularity' in df.columns:
    print(df['Cellularity'].value_counts(dropna=False))
print()

# Pam50: should be LumA, LumB, Her2, claudin-low, Basal, Normal, NC
print("Pam50 + Claudin-low subtype values:")
if 'Pam50 + Claudin-low subtype' in df.columns:
    print(df['Pam50 + Claudin-low subtype'].value_counts(dropna=False))
print()

# Tumor Stage: should be 0.0-4.0
print("Tumor Stage values:")
if 'Tumor Stage' in df.columns:
    print(df['Tumor Stage'].value_counts(dropna=False).sort_index())
print()

# Tumor Size: should be 1-182 mm range (note: column might be 'Tumor Size' or 'Tumor Size(mm)')
size_cols = [c for c in df.columns if 'tumor size' in c.lower() or 'tumor_size' in c.lower()]
print(f"Tumor size columns found: {size_cols}")
for col in size_cols:
    print(f"  {col}: min={df[col].min()}, max={df[col].max()}, median={df[col].median()}")
print()

# Relapse Free Status (the target): should be Recurred / Not Recurred (or similar)
print("Relapse Free Status (target) values:")
target_cols = [c for c in df.columns if 'relapse' in c.lower() and 'month' not in c.lower()]
print(f"Possible target columns: {target_cols}")
for col in target_cols:
    print(f"\n  {col}:")
    print(df[col].value_counts(dropna=False))

ER Status values:
ER Status
Positive    1825
Negative     644
NaN           40
Name: count, dtype: int64

ER status measured by IHC values:
ER status measured by IHC
Positve     1817
Negative     609
NaN           83
Name: count, dtype: int64

Cellularity values:
Cellularity
High        965
Moderate    737
NaN         592
Low         215
Name: count, dtype: int64

Pam50 + Claudin-low subtype values:
Pam50 + Claudin-low subtype
LumA           700
NaN            529
LumB           475
Her2           224
claudin-low    218
Basal          209
Normal         148
NC               6
Name: count, dtype: int64

Tumor Stage values:
Tumor Stage
0.0     24
1.0    630
2.0    979
3.0    144
4.0     11
NaN    721
Name: count, dtype: int64

Tumor size columns found: ['Tumor Size(mm)']
  Tumor Size(mm): min=1.0, max=182.0, median=22.41

Relapse Free Status (target) values:
Possible target columns: ['Relapse Free Status']

  Relapse Free Status:
Relapse Free Status
Not Recurred    1486
Recurred        1

In [6]:
# PDF says NPI = 0.2 × S(cm) + N + G
# So: NPI = 0.2 × (Tumor Size in mm / 10) + Lymph Nodes + Grade
# Let's see if the stored NPI matches the formula

if all(c in df.columns for c in ['Nottingham prognostic index', 'Lymph nodes examined positive']):
    # Find the right tumor size column
    size_col = None
    for c in ['Tumor Size', 'Tumor Size(mm)', 'Tumor_Size']:
        if c in df.columns:
            size_col = c
            break
    
    grade_col = None
    for c in ['Neoplasm Histologic Grade', 'G-Neoplasm Histologic Grade']:
        if c in df.columns:
            grade_col = c
            break
    
    if size_col and grade_col:
        complete = df.dropna(subset=[size_col, 'Lymph nodes examined positive', grade_col, 'Nottingham prognostic index'])
        computed_npi = 0.2 * (complete[size_col] / 10) + complete['Lymph nodes examined positive'] + complete[grade_col]
        diff = (complete['Nottingham prognostic index'] - computed_npi).abs()
        print(f"NPI formula check (using {size_col} and {grade_col}):")
        print(f"  Rows compared: {len(complete)}")
        print(f"  Mean absolute difference: {diff.mean():.4f}")
        print(f"  Max absolute difference: {diff.max():.4f}")
        print(f"  Rows where stored NPI matches formula (within 0.01): {(diff < 0.01).sum()} / {len(complete)}")
    else:
        print(f"Couldn't find size or grade column. Available: size={size_col}, grade={grade_col}")
else:
    print("NPI or lymph node column missing")

NPI formula check (using Tumor Size(mm) and G-Neoplasm Histologic Grade):
  Rows compared: 2040
  Mean absolute difference: 1.6856
  Max absolute difference: 42.4500
  Rows where stored NPI matches formula (within 0.01): 10 / 2040


In [7]:
# Test the 529-row missingness block hypothesis
block_features = [
    'Type of Breast Surgery', 'Chemotherapy', 'Pam50 + Claudin-low subtype',
    'HER2 Status', 'Hormone Therapy', 'Inferred Menopausal State',
    'Integrative Cluster', 'HER2 status measured by SNP6',
    'PR Status', 'Radio Therapy'
]

# Note: Type of Breast Surgery has 554 missing, slightly more than 529.
# Include it but the count differs slightly.

# For each feature, get the row indices that are missing
missing_indices = {f: set(df[df[f].isna()].index) for f in block_features}

# Print missing count per feature
print("Missing counts:")
for f, idx in missing_indices.items():
    print(f"  {f}: {len(idx)}")

# Find the intersection of features with exactly 529 missing
core_block_features = [f for f in block_features if len(missing_indices[f]) == 529]
print(f"\nFeatures with exactly 529 missing: {len(core_block_features)}")

if core_block_features:
    # How many rows are missing across ALL of these features?
    common_missing = set.intersection(*[missing_indices[f] for f in core_block_features])
    print(f"Rows missing all {len(core_block_features)} features simultaneously: {len(common_missing)}")
    
    # How many rows are missing at least one?
    union_missing = set.union(*[missing_indices[f] for f in core_block_features])
    print(f"Rows missing at least one of these features: {len(union_missing)}")
    
    # The decision rule from the plan
    overlap_pct = 100 * len(common_missing) / len(union_missing) if union_missing else 0
    print(f"\nOverlap: {overlap_pct:.1f}% (rule: ≥90% means treat as a block)")
    
    # Also check: do those 529 patients have a recognizable pattern?
    # e.g., all from a certain Cohort or with a certain ID prefix?
    if len(common_missing) > 0:
        block_df = df.loc[list(common_missing)]
        print(f"\nCohort distribution of the missing-block patients:")
        print(block_df['Cohort'].value_counts(dropna=False))
        print(f"\nID prefix of missing-block patients:")
        prefixes = block_df['Patient ID'].str.split('-').str[0].value_counts()
        print(prefixes)

Missing counts:
  Type of Breast Surgery: 554
  Chemotherapy: 529
  Pam50 + Claudin-low subtype: 529
  HER2 Status: 529
  Hormone Therapy: 529
  Inferred Menopausal State: 529
  Integrative Cluster: 529
  HER2 status measured by SNP6: 529
  PR Status: 529
  Radio Therapy: 529

Features with exactly 529 missing: 9
Rows missing all 9 features simultaneously: 529
Rows missing at least one of these features: 529

Overlap: 100.0% (rule: ≥90% means treat as a block)

Cohort distribution of the missing-block patients:
Cohort
1.0    288
7.0    105
8.0     82
9.0     40
NaN     11
6.0      3
Name: count, dtype: int64

ID prefix of missing-block patients:
Patient ID
MTS    524
MB       5
Name: count, dtype: int64


In [8]:
# === Step 1c (revised): Drop Tumor Stage, finalize column drops ===
# Documented decision: Tumor Stage excluded due to 26% missingness in main cohort.
# Mode imputation on a clinical staging variable is methodologically unsound;
# listwise deletion costs ~500 patients; the chapter's pipeline did not include
# Tumor Stage in its final feature set. Drop it.

cols_to_drop = [
    # Identifiers and constants
    'Patient ID', 'Cohort', 'Sex', 'Cancer Type',
    # Redundant encodings
    'Cancer Type Detailed', 'ER status measured by IHC',
    'HER2 status measured by SNP6', '3-Gene classifier subtype',
    'Oncotree Code', 'Integrative Cluster',
    # Pre-derived columns (logic isn't documented)
    'TNBC', 'NPI-Prognosis', 'Age40', 'S-Tumor-Size',
    'Survival Status', 'RecurredNotRecurredStatus',
    # Outcome-leaky columns
    'Relapse Free Status (Months)', 'Overall Survival (Months)',
    'Overall Survival Status', "Patient's Vital Status",
    # High missingness, methodologically problematic to impute
    'Tumor Stage',
]

df_predictors = df.drop(columns=cols_to_drop)
print(f"After dropping unused columns: {df_predictors.shape}")
print(f"Predictor count: {df_predictors.shape[1] - 1}")  # minus the target
print(f"Remaining columns: {df_predictors.columns.tolist()}")

After dropping unused columns: (2509, 19)
Predictor count: 18
Remaining columns: ['Age at Diagnosis', 'Type of Breast Surgery', 'Cellularity', 'Chemotherapy', 'Pam50 + Claudin-low subtype', 'ER Status', 'G-Neoplasm Histologic Grade', 'HER2 Status', 'Tumor Other Histologic Subtype', 'Hormone Therapy', 'Inferred Menopausal State', 'Primary Tumor Laterality', 'Lymph nodes examined positive', 'Mutation Count', 'Nottingham prognostic index', 'PR Status', 'Radio Therapy', 'Relapse Free Status', 'Tumor Size(mm)']


In [9]:
# === Step 1d: Universal drops (apply to BOTH cleaning regimes) ===

# 1. Drop the 529-patient MTS-T block (identified by missingness on any block feature)
mts_block_mask = df_predictors['Chemotherapy'].isna()

# 2. Drop the 3 Breast Sarcoma patients (different disease entity)
sarcoma_mask = df['Cancer Type'] == 'Breast Sarcoma'

# 3. Drop patients with missing target
missing_target_mask = df_predictors['Relapse Free Status'].isna()

# 4. Drop the 6 Pam50 = NC patients (too small a category for analysis)
nc_mask = df_predictors['Pam50 + Claudin-low subtype'] == 'NC'

# Combine
to_drop = mts_block_mask | sarcoma_mask | missing_target_mask | nc_mask
print(f"MTS block patients to drop: {mts_block_mask.sum()}")
print(f"Sarcoma patients to drop: {sarcoma_mask.sum()}")
print(f"Missing target patients to drop: {missing_target_mask.sum()}")
print(f"Pam50=NC patients to drop: {nc_mask.sum()}")
print(f"\nTotal rows to drop (universal): {to_drop.sum()}")

df_universal = df_predictors[~to_drop].copy()
print(f"\nShape after universal drops: {df_universal.shape}")

# Show remaining missingness
print(f"\nRemaining missingness in main cohort:")
remaining_missing = df_universal.isna().sum().sort_values(ascending=False)
remaining_missing = remaining_missing[remaining_missing > 0]
remaining_pct = (remaining_missing / len(df_universal) * 100).round(2)
remaining_df = pd.DataFrame({'count': remaining_missing, 'pct': remaining_pct})
print(remaining_df.to_string())

MTS block patients to drop: 529
Sarcoma patients to drop: 3
Missing target patients to drop: 21
Pam50=NC patients to drop: 6

Total rows to drop (universal): 539

Shape after universal drops: (1970, 19)

Remaining missingness in main cohort:
                                count   pct
Mutation Count                    119  6.04
Primary Tumor Laterality          108  5.48
G-Neoplasm Histologic Grade        84  4.26
Lymph nodes examined positive      74  3.76
Cellularity                        62  3.15
Tumor Other Histologic Subtype     42  2.13
Type of Breast Surgery             24  1.22
Tumor Size(mm)                     23  1.17


In [10]:
# === Step 1f-prep: Diagnostic checks before producing the two cleaned datasets ===

# Check 1: Class balance after universal drops
print("Class balance after universal drops:")
target_balance = df_universal['Relapse Free Status'].value_counts()
target_pct = df_universal['Relapse Free Status'].value_counts(normalize=True) * 100
print(pd.DataFrame({'count': target_balance, 'pct': target_pct.round(2)}).to_string())
print()

# Check 2: ER+/no-hormone-therapy cell size (semantic loss viability)
er_pos_no_ht = df_universal[
    (df_universal['ER Status'] == 'Positive') &
    (df_universal['Hormone Therapy'] == 'No')
]
print(f"ER+ / no-hormone-therapy cell size: {len(er_pos_no_ht)}")
print(f"  Of those, recurred: {(er_pos_no_ht['Relapse Free Status'] == 'Recurred').sum()}")
print(f"  Of those, not recurred: {(er_pos_no_ht['Relapse Free Status'] == 'Not Recurred').sum()}")
print(f"  Recurrence rate in this subgroup: "
      f"{(er_pos_no_ht['Relapse Free Status'] == 'Recurred').mean() * 100:.1f}%")
print(f"  Recurrence rate overall: "
      f"{(df_universal['Relapse Free Status'] == 'Recurred').mean() * 100:.1f}%")
print()

# Check 3: Pam50 distribution (after dropping NC)
print("Pam50 + Claudin-low subtype distribution:")
print(df_universal['Pam50 + Claudin-low subtype'].value_counts(dropna=False).to_string())
print()

# Check 4: Continuous feature ranges
print("Continuous feature summary:")
continuous = ['Age at Diagnosis', 'Lymph nodes examined positive',
              'Mutation Count', 'Nottingham prognostic index', 'Tumor Size(mm)',
              'G-Neoplasm Histologic Grade']
print(df_universal[continuous].describe().T.to_string())

Class balance after universal drops:
                     count    pct
Relapse Free Status              
Not Recurred          1170  59.39
Recurred               800  40.61

ER+ / no-hormone-therapy cell size: 423
  Of those, recurred: 171
  Of those, not recurred: 252
  Recurrence rate in this subgroup: 40.4%
  Recurrence rate overall: 40.6%

Pam50 + Claudin-low subtype distribution:
Pam50 + Claudin-low subtype
LumA           699
LumB           475
Her2           224
claudin-low    215
Basal          209
Normal         148

Continuous feature summary:
                                count       mean        std    min     25%     50%       75%     max
Age at Diagnosis               1970.0  61.062081  12.963961  21.93  51.365  61.785  70.58000   96.29
Lymph nodes examined positive  1896.0   2.004219   4.085108   0.00   0.000   0.000   2.00000   45.00
Mutation Count                 1851.0   5.690978   4.058973   1.00   3.000   5.000   7.00000   80.00
Nottingham prognostic index    1970.0

In [11]:
# === Step 1f: Produce the two cleaned datasets ===

# === Permissive cleaning ===
# Median impute continuous, mode impute categorical
df_permissive = df_universal.copy()

continuous_cols = [
    'Mutation Count', 'Lymph nodes examined positive', 'Tumor Size(mm)',
    'G-Neoplasm Histologic Grade'
]
categorical_cols_to_impute = [
    'Cellularity', 'Primary Tumor Laterality',
    'Tumor Other Histologic Subtype', 'Type of Breast Surgery'
]

# Track how many we're imputing per column
imputation_log = {}
for col in continuous_cols:
    n_missing = df_permissive[col].isna().sum()
    median_val = df_permissive[col].median()
    df_permissive[col] = df_permissive[col].fillna(median_val)
    imputation_log[col] = (n_missing, f"median = {median_val}")

for col in categorical_cols_to_impute:
    n_missing = df_permissive[col].isna().sum()
    mode_val = df_permissive[col].mode()[0]
    df_permissive[col] = df_permissive[col].fillna(mode_val)
    imputation_log[col] = (n_missing, f"mode = {mode_val!r}")

# NPI is calculated, but missing in 222 rows from the original. Recompute or median-impute?
# Since NPI didn't match the formula in cell 6, treat it as a stored value and median-impute.
n_missing_npi = df_permissive['Nottingham prognostic index'].isna().sum()
median_npi = df_permissive['Nottingham prognostic index'].median()
df_permissive['Nottingham prognostic index'] = df_permissive['Nottingham prognostic index'].fillna(median_npi)
imputation_log['Nottingham prognostic index'] = (n_missing_npi, f"median = {median_npi}")

print("Imputation log:")
for col, (n, val) in imputation_log.items():
    print(f"  {col}: {n} values imputed with {val}")

# Verify no missingness remains
permissive_missing = df_permissive.isna().sum().sum()
print(f"\nPermissive dataset shape: {df_permissive.shape}")
print(f"Total missing values remaining: {permissive_missing}")
assert permissive_missing == 0, "Permissive cleaning still has missing values!"

# === Strict cleaning ===
# Listwise drop any row with any missing value
df_strict = df_universal.dropna().copy()
print(f"\nStrict dataset shape: {df_strict.shape}")
print(f"Strict cleaning dropped {len(df_universal) - len(df_strict)} additional rows")

# Sanity check class balance in strict
print(f"\nStrict cleaning class balance:")
strict_balance = df_strict['Relapse Free Status'].value_counts(normalize=True) * 100
print(strict_balance.round(2).to_string())

Imputation log:
  Mutation Count: 119 values imputed with median = 5.0
  Lymph nodes examined positive: 74 values imputed with median = 0.0
  Tumor Size(mm): 23 values imputed with median = 23.0
  G-Neoplasm Histologic Grade: 84 values imputed with median = 3.0
  Cellularity: 62 values imputed with mode = 'High'
  Primary Tumor Laterality: 108 values imputed with mode = 'Left'
  Tumor Other Histologic Subtype: 42 values imputed with mode = 'Ductal/NST'
  Type of Breast Surgery: 24 values imputed with mode = 'Mastectomy'
  Nottingham prognostic index: 0 values imputed with median = 4.042

Permissive dataset shape: (1970, 19)
Total missing values remaining: 0

Strict dataset shape: (1603, 19)
Strict cleaning dropped 367 additional rows

Strict cleaning class balance:
Relapse Free Status
Not Recurred    59.39
Recurred        40.61


In [12]:
# === Save outputs ===
from pathlib import Path

OUTPUT_DIR = Path('../data/processed')
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

df_permissive.to_csv(OUTPUT_DIR / 'metabric_permissive.csv', index=False)
df_strict.to_csv(OUTPUT_DIR / 'metabric_strict.csv', index=False)

print(f"Saved metabric_permissive.csv: {df_permissive.shape}")
print(f"Saved metabric_strict.csv: {df_strict.shape}")

Saved metabric_permissive.csv: (1970, 19)
Saved metabric_strict.csv: (1603, 19)


In [13]:
# === Step 1g-prep: Generate and save train/test split indices ===
from sklearn.model_selection import train_test_split

# Use permissive cleaning as the basis for the canonical split.
# We need to map split indices back to original row positions so they work
# across all encoded variants of the same dataset.

# Reset index so we have clean 0-based indexing
df_permissive_reset = df_permissive.reset_index(drop=True)
df_strict_reset = df_strict.reset_index(drop=True)

# Permissive split (canonical)
X_perm = df_permissive_reset.drop(columns=['Relapse Free Status'])
y_perm = df_permissive_reset['Relapse Free Status']

train_idx_perm, test_idx_perm = train_test_split(
    df_permissive_reset.index,
    test_size=0.2,
    random_state=42,
    stratify=y_perm,
)

# Strict split (separate, since strict has different rows)
X_strict = df_strict_reset.drop(columns=['Relapse Free Status'])
y_strict = df_strict_reset['Relapse Free Status']

train_idx_strict, test_idx_strict = train_test_split(
    df_strict_reset.index,
    test_size=0.2,
    random_state=42,
    stratify=y_strict,
)

# Save split indices for both regimes
splits = pd.DataFrame({
    'permissive_train': pd.Series(sorted(train_idx_perm)),
    'permissive_test': pd.Series(sorted(test_idx_perm)),
    'strict_train': pd.Series(sorted(train_idx_strict)),
    'strict_test': pd.Series(sorted(test_idx_strict)),
})
splits.to_csv(OUTPUT_DIR / 'split_indices.csv', index=False)

print(f"Permissive: {len(train_idx_perm)} train, {len(test_idx_perm)} test")
print(f"Strict: {len(train_idx_strict)} train, {len(test_idx_strict)} test")

# Verify class balance in train and test
print(f"\nPermissive train class balance:")
print(y_perm.loc[train_idx_perm].value_counts(normalize=True).round(3))
print(f"\nPermissive test class balance:")
print(y_perm.loc[test_idx_perm].value_counts(normalize=True).round(3))

# Save the reset-index versions (so split indices match)
df_permissive_reset.to_csv(OUTPUT_DIR / 'metabric_permissive.csv', index=False)
df_strict_reset.to_csv(OUTPUT_DIR / 'metabric_strict.csv', index=False)
print("\nResaved permissive and strict CSVs with reset index.")

Permissive: 1576 train, 394 test
Strict: 1282 train, 321 test

Permissive train class balance:
Relapse Free Status
Not Recurred    0.594
Recurred        0.406
Name: proportion, dtype: float64

Permissive test class balance:
Relapse Free Status
Not Recurred    0.594
Recurred        0.406
Name: proportion, dtype: float64

Resaved permissive and strict CSVs with reset index.


In [14]:
# === Step 1g (a): Produce the encoded variant for MLP/SHAP/numeric methods ===
# - One-hot encode low-cardinality categoricals
# - Label-encode ordinals (preserving order)
# - Standardize continuous features (z-score)
# - Map target to 0/1

from sklearn.preprocessing import StandardScaler

def make_encoded(df_in, fit_scaler_on=None):
    """
    Encode a permissive- or strict-cleaned dataframe into all-numeric form.
    
    If fit_scaler_on is provided (a fitted StandardScaler), use it to transform.
    Otherwise fit a new one on df_in and return it.
    Returns (encoded_df, scaler).
    """
    df_out = df_in.copy()
    
    # Target: map to 0/1
    df_out['Relapse Free Status'] = df_out['Relapse Free Status'].map({
        'Not Recurred': 0, 'Recurred': 1
    })
    
    # Ordinal: Cellularity (Low < Moderate < High)
    cellularity_order = {'Low': 0, 'Moderate': 1, 'High': 2}
    df_out['Cellularity'] = df_out['Cellularity'].map(cellularity_order)
    
    # G-Neoplasm Histologic Grade is already numeric (1, 2, 3) — leave as-is
    # Tumor Stage is dropped, so no concern there
    
    # Low-cardinality categoricals — one-hot encode
    one_hot_cols = [
        'Type of Breast Surgery', 'Chemotherapy', 'Pam50 + Claudin-low subtype',
        'ER Status', 'HER2 Status', 'Tumor Other Histologic Subtype',
        'Hormone Therapy', 'Inferred Menopausal State',
        'Primary Tumor Laterality', 'PR Status', 'Radio Therapy',
    ]
    df_out = pd.get_dummies(df_out, columns=one_hot_cols, drop_first=False)
    
    # Convert any remaining bool columns from one-hot to int
    bool_cols = df_out.select_dtypes(include=['bool']).columns
    df_out[bool_cols] = df_out[bool_cols].astype(int)
    
    # Continuous features — standardize
    continuous_cols = [
        'Age at Diagnosis', 'Lymph nodes examined positive',
        'Mutation Count', 'Nottingham prognostic index',
        'Tumor Size(mm)', 'G-Neoplasm Histologic Grade', 'Cellularity'
    ]
    
    if fit_scaler_on is None:
        scaler = StandardScaler()
        df_out[continuous_cols] = scaler.fit_transform(df_out[continuous_cols])
    else:
        scaler = fit_scaler_on
        df_out[continuous_cols] = scaler.transform(df_out[continuous_cols])
    
    return df_out, scaler


# Encode permissive (fit scaler on training portion only to avoid leakage)
df_perm_encoded, perm_scaler = make_encoded(df_permissive_reset)

# Critical correctness check: scaler should be fit on TRAIN only.
# The line above fits on the whole dataset, which is a small leakage.
# Refit properly:
df_perm_train_for_fit = df_permissive_reset.loc[train_idx_perm].copy()
_, perm_scaler = make_encoded(df_perm_train_for_fit)
df_perm_encoded, _ = make_encoded(df_permissive_reset, fit_scaler_on=perm_scaler)

# Same for strict
df_strict_train_for_fit = df_strict_reset.loc[train_idx_strict].copy()
_, strict_scaler = make_encoded(df_strict_train_for_fit)
df_strict_encoded, _ = make_encoded(df_strict_reset, fit_scaler_on=strict_scaler)

# Save
df_perm_encoded.to_csv(OUTPUT_DIR / 'metabric_permissive_encoded.csv', index=False)
df_strict_encoded.to_csv(OUTPUT_DIR / 'metabric_strict_encoded.csv', index=False)

print(f"Permissive encoded shape: {df_perm_encoded.shape}")
print(f"Strict encoded shape: {df_strict_encoded.shape}")
print(f"\nFirst few columns:")
print(df_perm_encoded.columns[:15].tolist())
print(f"\nDtype check (should all be numeric):")
print(df_perm_encoded.dtypes.value_counts())

Permissive encoded shape: (1970, 40)
Strict encoded shape: (1603, 39)

First few columns:
['Age at Diagnosis', 'Cellularity', 'G-Neoplasm Histologic Grade', 'Lymph nodes examined positive', 'Mutation Count', 'Nottingham prognostic index', 'Relapse Free Status', 'Tumor Size(mm)', 'Type of Breast Surgery_Breast Conserving', 'Type of Breast Surgery_Mastectomy', 'Chemotherapy_No', 'Chemotherapy_Yes', 'Pam50 + Claudin-low subtype_Basal', 'Pam50 + Claudin-low subtype_Her2', 'Pam50 + Claudin-low subtype_LumA']

Dtype check (should all be numeric):
int64      33
float64     7
Name: count, dtype: int64


In [15]:
# === Step 1g (b): Produce the binarized variant for GOSDT ===
# All features become binary indicators.
# Continuous: threshold at multiple cutpoints (we'll use quartiles + the median)
# Ordinal: threshold at each level
# Categorical: already-binary indicators from one-hot

def make_binarized(df_in, fit_thresholds_on=None):
    """
    Binarize a cleaned dataframe for GOSDT.
    
    For continuous features, threshold at multiple cutpoints.
    For ordinal/categorical, produce indicator columns.
    
    If fit_thresholds_on is provided (a dict of column -> list of cutpoints),
    use those. Otherwise compute from df_in and return them.
    """
    df_out = pd.DataFrame(index=df_in.index)
    
    # Target — keep as 0/1
    df_out['target'] = df_in['Relapse Free Status'].map({
        'Not Recurred': 0, 'Recurred': 1
    })
    
    thresholds = {} if fit_thresholds_on is None else dict(fit_thresholds_on)
    
    # Continuous features — threshold at quartiles
    continuous_cols = [
        'Age at Diagnosis', 'Lymph nodes examined positive',
        'Mutation Count', 'Nottingham prognostic index',
        'Tumor Size(mm)',
    ]
    
    for col in continuous_cols:
        if col in thresholds:
            cutpoints = thresholds[col]
        else:
            # Use quartiles as cutpoints (gives 3 binary indicators per feature)
            cutpoints = [
                df_in[col].quantile(0.25),
                df_in[col].quantile(0.50),
                df_in[col].quantile(0.75),
            ]
            # Deduplicate in case of ties (e.g., lymph nodes has many zeros)
            cutpoints = sorted(set(cutpoints))
            thresholds[col] = cutpoints
        
        for cp in cutpoints:
            df_out[f"{col} >= {cp:.2f}"] = (df_in[col] >= cp).astype(int)
    
    # Ordinal: Histologic Grade (1, 2, 3) — threshold at each level
    df_out["Grade >= 2"] = (df_in['G-Neoplasm Histologic Grade'] >= 2).astype(int)
    df_out["Grade >= 3"] = (df_in['G-Neoplasm Histologic Grade'] >= 3).astype(int)
    
    # Ordinal: Cellularity (Low=0, Moderate=1, High=2) — threshold at each level
    cellularity_order = {'Low': 0, 'Moderate': 1, 'High': 2}
    cellularity_numeric = df_in['Cellularity'].map(cellularity_order)
    df_out["Cellularity >= Moderate"] = (cellularity_numeric >= 1).astype(int)
    df_out["Cellularity >= High"] = (cellularity_numeric >= 2).astype(int)
    
    # Binary categoricals (Yes/No, Positive/Negative, etc.)
    binary_mappings = {
        'Type of Breast Surgery': {'Mastectomy': 1, 'Breast Conserving': 0},
        'Chemotherapy': {'Yes': 1, 'No': 0},
        'ER Status': {'Positive': 1, 'Negative': 0},
        'HER2 Status': {'Positive': 1, 'Negative': 0},
        'Hormone Therapy': {'Yes': 1, 'No': 0},
        'Inferred Menopausal State': {'Post': 1, 'Pre': 0},
        'Primary Tumor Laterality': {'Left': 1, 'Right': 0},
        'PR Status': {'Positive': 1, 'Negative': 0},
        'Radio Therapy': {'Yes': 1, 'No': 0},
    }
    for col, mapping in binary_mappings.items():
        df_out[col] = df_in[col].map(mapping).astype(int)
    
    # Multi-class categoricals — one-hot
    multiclass_cols = [
        'Pam50 + Claudin-low subtype',
        'Tumor Other Histologic Subtype',
    ]
    for col in multiclass_cols:
        dummies = pd.get_dummies(df_in[col], prefix=col).astype(int)
        df_out = pd.concat([df_out, dummies], axis=1)
    
    return df_out, thresholds


# Fit thresholds on training data only
df_perm_train_for_fit = df_permissive_reset.loc[train_idx_perm].copy()
_, perm_thresholds = make_binarized(df_perm_train_for_fit)
df_perm_binarized, _ = make_binarized(df_permissive_reset, fit_thresholds_on=perm_thresholds)

df_strict_train_for_fit = df_strict_reset.loc[train_idx_strict].copy()
_, strict_thresholds = make_binarized(df_strict_train_for_fit)
df_strict_binarized, _ = make_binarized(df_strict_reset, fit_thresholds_on=strict_thresholds)

# Save
df_perm_binarized.to_csv(OUTPUT_DIR / 'metabric_permissive_binarized.csv', index=False)
df_strict_binarized.to_csv(OUTPUT_DIR / 'metabric_strict_binarized.csv', index=False)

print(f"Permissive binarized shape: {df_perm_binarized.shape}")
print(f"Strict binarized shape: {df_strict_binarized.shape}")
print(f"\nAll columns binary or 0/1 integer: "
      f"{(df_perm_binarized.dtypes.isin(['int64', 'int32'])).all()}")
print(f"\nFirst 20 columns:")
for c in df_perm_binarized.columns[:20]:
    print(f"  {c}")

# Show the thresholds used (for documentation)
print(f"\nThresholds used for continuous features (permissive):")
for col, cps in perm_thresholds.items():
    print(f"  {col}: {[round(cp, 2) for cp in cps]}")

Permissive binarized shape: (1970, 42)
Strict binarized shape: (1603, 41)

All columns binary or 0/1 integer: False

First 20 columns:
  target
  Age at Diagnosis >= 51.36
  Age at Diagnosis >= 61.59
  Age at Diagnosis >= 70.52
  Lymph nodes examined positive >= 0.00
  Lymph nodes examined positive >= 2.00
  Mutation Count >= 3.00
  Mutation Count >= 5.00
  Mutation Count >= 7.00
  Nottingham prognostic index >= 3.05
  Nottingham prognostic index >= 4.04
  Nottingham prognostic index >= 5.04
  Tumor Size(mm) >= 18.00
  Tumor Size(mm) >= 23.00
  Tumor Size(mm) >= 30.00
  Grade >= 2
  Grade >= 3
  Cellularity >= Moderate
  Cellularity >= High
  Type of Breast Surgery

Thresholds used for continuous features (permissive):
  Age at Diagnosis: [np.float64(51.36), np.float64(61.58), np.float64(70.52)]
  Lymph nodes examined positive: [np.float64(0.0), np.float64(2.0)]
  Mutation Count: [np.float64(3.0), np.float64(5.0), np.float64(7.0)]
  Nottingham prognostic index: [np.float64(3.05), np.fl

In [16]:
# === Step 1 verification: confirm everything saved and sanity check ===

import os

expected_files = [
    'metabric_permissive.csv',
    'metabric_strict.csv',
    'metabric_permissive_encoded.csv',
    'metabric_strict_encoded.csv',
    'metabric_permissive_binarized.csv',
    'metabric_strict_binarized.csv',
    'split_indices.csv',
]

print("Output files:")
for f in expected_files:
    path = OUTPUT_DIR / f
    if path.exists():
        size_kb = os.path.getsize(path) / 1024
        # Quick row count
        n_rows = sum(1 for _ in open(path)) - 1  # minus header
        print(f"  ✓ {f}: {n_rows} rows, {size_kb:.1f} KB")
    else:
        print(f"  ✗ {f}: MISSING")

# Sanity check: train/test split should round-trip cleanly
print("\nSplit verification:")
print(f"  Permissive train + test = {len(train_idx_perm) + len(test_idx_perm)} (should be {len(df_permissive_reset)})")
print(f"  Strict train + test = {len(train_idx_strict) + len(test_idx_strict)} (should be {len(df_strict_reset)})")
print(f"  Train/test overlap (permissive): {len(set(train_idx_perm) & set(test_idx_perm))} (should be 0)")
print(f"  Train/test overlap (strict): {len(set(train_idx_strict) & set(test_idx_strict))} (should be 0)")

Output files:
  ✓ metabric_permissive.csv: 1970 rows, 240.7 KB
  ✓ metabric_strict.csv: 1603 rows, 195.8 KB
  ✓ metabric_permissive_encoded.csv: 1970 rows, 396.2 KB
  ✓ metabric_strict_encoded.csv: 1603 rows, 319.7 KB
  ✓ metabric_permissive_binarized.csv: 1970 rows, 164.7 KB
  ✓ metabric_strict_binarized.csv: 1603 rows, 131.0 KB
  ✓ split_indices.csv: 1576 rows, 23.4 KB

Split verification:
  Permissive train + test = 1970 (should be 1970)
  Strict train + test = 1603 (should be 1603)
  Train/test overlap (permissive): 0 (should be 0)
  Train/test overlap (strict): 0 (should be 0)


In [17]:
# Diagnostic: what's the actual dtype distribution in the binarized variant?
print("Binarized dtype distribution:")
print(df_perm_binarized.dtypes.value_counts())
print()

# Check for non-integer columns
non_int = df_perm_binarized.select_dtypes(exclude=['int64', 'int32', 'int8'])
if len(non_int.columns) > 0:
    print(f"Non-integer columns in binarized:")
    for col in non_int.columns:
        print(f"  {col}: dtype={df_perm_binarized[col].dtype}, "
              f"unique values: {sorted(df_perm_binarized[col].unique())[:5]}")
else:
    print("All columns are integer.")
print()

# Check for constant columns (every value is the same)
print("Constant columns in binarized (no discriminative value):")
nunique = df_perm_binarized.nunique()
constant_cols = nunique[nunique == 1].index.tolist()
for col in constant_cols:
    val = df_perm_binarized[col].iloc[0]
    print(f"  {col}: every value is {val}")
if not constant_cols:
    print("  (none)")
print()

# Check column counts more carefully
print(f"Total binarized columns: {len(df_perm_binarized.columns)}")
print(f"All columns: {df_perm_binarized.columns.tolist()}")

Binarized dtype distribution:
int64    42
Name: count, dtype: int64

All columns are integer.

Constant columns in binarized (no discriminative value):
  Lymph nodes examined positive >= 0.00: every value is 1

Total binarized columns: 42
All columns: ['target', 'Age at Diagnosis >= 51.36', 'Age at Diagnosis >= 61.59', 'Age at Diagnosis >= 70.52', 'Lymph nodes examined positive >= 0.00', 'Lymph nodes examined positive >= 2.00', 'Mutation Count >= 3.00', 'Mutation Count >= 5.00', 'Mutation Count >= 7.00', 'Nottingham prognostic index >= 3.05', 'Nottingham prognostic index >= 4.04', 'Nottingham prognostic index >= 5.04', 'Tumor Size(mm) >= 18.00', 'Tumor Size(mm) >= 23.00', 'Tumor Size(mm) >= 30.00', 'Grade >= 2', 'Grade >= 3', 'Cellularity >= Moderate', 'Cellularity >= High', 'Type of Breast Surgery', 'Chemotherapy', 'ER Status', 'HER2 Status', 'Hormone Therapy', 'Inferred Menopausal State', 'Primary Tumor Laterality', 'PR Status', 'Radio Therapy', 'Pam50 + Claudin-low subtype_Basal', '

In [18]:
# === Patch: drop constant columns from binarized variants ===

def drop_constant_cols(df_in, name=""):
    nunique = df_in.nunique()
    constant_cols = nunique[nunique == 1].index.tolist()
    if constant_cols:
        print(f"  {name}: dropping {len(constant_cols)} constant columns: {constant_cols}")
    return df_in.drop(columns=constant_cols)

df_perm_binarized = drop_constant_cols(df_perm_binarized, "permissive")
df_strict_binarized = drop_constant_cols(df_strict_binarized, "strict")

# Resave
df_perm_binarized.to_csv(OUTPUT_DIR / 'metabric_permissive_binarized.csv', index=False)
df_strict_binarized.to_csv(OUTPUT_DIR / 'metabric_strict_binarized.csv', index=False)

print(f"\nFinal permissive binarized shape: {df_perm_binarized.shape}")
print(f"Final strict binarized shape: {df_strict_binarized.shape}")

# Document the cutpoints actually used (with the constant ones removed)
print(f"\nFinal cutpoints for continuous features:")
for col_prefix in ['Age at Diagnosis', 'Lymph nodes examined positive',
                   'Mutation Count', 'Nottingham prognostic index', 'Tumor Size(mm)']:
    matching = [c for c in df_perm_binarized.columns if c.startswith(col_prefix + ' >=')]
    print(f"  {col_prefix}: {len(matching)} indicator(s) — {matching}")

  permissive: dropping 1 constant columns: ['Lymph nodes examined positive >= 0.00']
  strict: dropping 1 constant columns: ['Lymph nodes examined positive >= 0.00']

Final permissive binarized shape: (1970, 41)
Final strict binarized shape: (1603, 40)

Final cutpoints for continuous features:
  Age at Diagnosis: 3 indicator(s) — ['Age at Diagnosis >= 51.36', 'Age at Diagnosis >= 61.59', 'Age at Diagnosis >= 70.52']
  Lymph nodes examined positive: 1 indicator(s) — ['Lymph nodes examined positive >= 2.00']
  Mutation Count: 3 indicator(s) — ['Mutation Count >= 3.00', 'Mutation Count >= 5.00', 'Mutation Count >= 7.00']
  Nottingham prognostic index: 3 indicator(s) — ['Nottingham prognostic index >= 3.05', 'Nottingham prognostic index >= 4.04', 'Nottingham prognostic index >= 5.04']
  Tumor Size(mm): 3 indicator(s) — ['Tumor Size(mm) >= 18.00', 'Tumor Size(mm) >= 23.00', 'Tumor Size(mm) >= 30.00']


## Phase 1 Summary

Starting from 2,509 patients × 40 columns in the raw METABRIC file, applied:

**Universal exclusions** (applied to all downstream regimes):
- 529-patient MTS-T cohort (missing 9 features simultaneously, documented sub-study with incomplete clinical phenotyping)
- 3 Breast Sarcoma patients (different disease entity)
- 21 patients with missing recurrence target
- 6 patients with Pam50 = NC (category too small)

**Column exclusions:**
- Identifiers, constants, redundant encodings, pre-derived columns, outcome-leaky columns
- Tumor Stage (26% missingness in main cohort, no honest imputation; chapter's pipeline did not use it in final model)

**Cleaning regimes:**
- Permissive: median/mode imputation on residual missingness → 1,970 patients
- Strict: listwise deletion → 1,603 patients

**Outputs:**
- `metabric_permissive.csv`, `metabric_strict.csv` (tidy form)
- `metabric_permissive_encoded.csv`, `metabric_strict_encoded.csv` (one-hot + label-encoded + standardized for MLP/SHAP)
- `metabric_permissive_binarized.csv`, `metabric_strict_binarized.csv` (quartile-binarized for GOSDT)
- `split_indices.csv` (canonical 80/20 train/test split, stratified by target)

**Notable findings:**
- ER+/no-hormone-therapy subgroup recurrence rate (40.4%) equals overall rate (40.6%) — see Phase 6 for reframing implication
- NPI does not match its documented formula on this data (only 10/2,040 rows match within 0.01) — treated as stored value

In [19]:
# Pattern for downstream notebooks:
splits = pd.read_csv('../data/processed/split_indices.csv')
train_idx = splits['permissive_train'].dropna().astype(int).values
test_idx = splits['permissive_test'].dropna().astype(int).values

df = pd.read_csv('../data/processed/metabric_permissive.csv')
X = df.drop(columns=['Relapse Free Status'])
y = df['Relapse Free Status'].map({'Not Recurred': 0, 'Recurred': 1})

X_train, y_train = X.iloc[train_idx], y.iloc[train_idx]
X_test, y_test = X.iloc[test_idx], y.iloc[test_idx]